[CORE-7750] Creating topic with huge number of partitions leads to segfault

IoannisRP opened the pull request:
Fixes: [CORE-7750](https://redpandadata.atlassian.net/browse/CORE-7750)

The size of the `allocation_request` created as part of the partition allocation scales linearly with the number of partitions needed to be created. Currently, the check to see if the request can be accepted happens on the `allocation_request` object. When the number of requested partitions is too big, we run out of memory while creating this object.

The fix is to be lazy on the creation of the `allocation_request`, where possible, and do the first step of the validation before we create the full object.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v24.3.x
- [x] v24.2.x
- [ ] v24.1.x

## Release Notes

### Bug Fixes

* Fixed an issue where creating a topic with a huge number of partitions could lead to a crash.


[CORE-7750]: https://redpandadata.atlassian.net/browse/CORE-7750?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

rockwotj commented on src/v/cluster/scheduling/partition_allocator.h:
```diff
@@ -131,7 +138,10 @@ class partition_allocator {
 private:
     std::error_code
-    check_cluster_limits(const allocation_request& request) const;
+    check_cluster_limits(const uint64_t new_partitions_requested) const;
```
Can we call this `new_partition_replicas_requested` and add a small comment to the method about this being replication factor * partitions.

rockwotj commented on src/v/cluster/scheduling/types.h:
```diff
@@ -360,4 +360,25 @@ struct allocation_request {
     friend std::ostream& operator<<(std::ostream&, const allocation_request&);
 };
 
+struct raw_allocation_request {
```
Can we please add comments to both of the request structures and mention why one should use one or the other? So someone doesn't try to "cleanup" the code later and they miss the memory blowup.

rockwotj commented on src/v/cluster/scheduling/partition_allocator.cc:
```diff
@@ -240,18 +232,58 @@ std::error_code partition_allocator::check_cluster_limits(
+    const uint64_t create_count = static_cast<uint64_t>(raw.partition_count)
+                                  * static_cast<uint64_t>(
+                                    raw.replication_factor);
+    auto cluster_errc = check_cluster_limits(create_count);
```
Worth adding a comment that we must check limits before adding as to not materialize large vectors.

rockwotj commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -481,26 +481,34 @@ make_error_result(const model::topic_namespace& tp_ns, std::error_code ec) {
 }
 
-static allocation_request make_allocation_request(
+static raw_allocation_request make_raw_allocation_request(
```
nit: prefer anonymous namespaces to static methods.

rockwotj reviewed (commented):
Looks great to me generally thanks! Just some small nits on docs/naming.

IoannisRP commented:
@rockwotj good points. I'll address them and i'll add the `do_create_partion` lazy path.

michael-redpanda commented:
/dt

vbotbuildovich commented:
<!--UNIT FAILURES-->
<!--3892e6b4-0beb-4bf8-87b6-a042da8ea366-->
<!--body-->
the below tests from https://buildkite.com/redpanda/redpanda/builds/58192#01933fcb-6e85-465d-9b8e-7ea931b6e484 have failed and will be retried
```
translator_test_rpfixture
```
<!--/body-->

dotnwat commented:
seems like this is primarily a replication team code area. added reviewers from that team

IoannisRP commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -802,8 +809,12 @@ ss::future<topic_result> topics_frontend::do_create_topic(
         partition_allocator& al) {
+          if (assignable_config.has_custom_assignment()) {
+              return al.allocate(
+                make_custom_allocation_request(assignable_config, topic_aware));
```
Mainly, we cannot be lazy about these paths, as the request has per-partition requirements.

Maybe we could improve this by delaying the transform, but I don't think there would be any obvious benefits.
The raw transformation takes an integer and instantiates an array of that size. 
This path translates an array into another array.
While it is very easy to request for a "billion" new partitions, it is not as easy to request for a billion custom assignments. If someone were to indeed ask for that, the point of failure would probably be around the parsing of the request and the population of this array. It would have to be handled there.

mmaslankaprv commented on src/v/cluster/scheduling/types.h:
```diff
@@ -360,4 +360,31 @@ struct allocation_request {
+ *  The memory required for a raw_allocation_request objects does not scale
+ *  with the number of requested partitions.
+ */
+struct raw_allocation_request {
```
nit: maybe we can name that `simple_allocation_request` ?
Also consider separating a commit in which a request is added from the commit in which it is used

IoannisRP commented:
- Renamed `raw_allocation_request` to `simple_allocation_request`
- Renamed `partition_count` -> `additional_partitions`
- Removed single argument constructor
- Split implementation of new path into 2 separate commits.

BenPope commented on src/v/cluster/tests/partition_allocator_tests.cc:
```diff
@@ -341,6 +381,27 @@ FIXTURE_TEST(allocation_units_test, partition_allocator_fixture) {
     // we do not decrement the highest raft group
     BOOST_REQUIRE_EQUAL(allocator().state().last_group_id()(), 10);
 }
+FIXTURE_TEST(allocation_units_test_raw_req, partition_allocator_fixture) {
```
nitpick: `raw` vs `simple`

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
Failed to create a backport PR to v24.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-24135-v24.2.x-772 remotes/upstream/v24.2.x
git cherry-pick -x ef6a6ac0ee fc1047866e 16a95fb467
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/11951712861).
